zerver/tests/test_channel_permissions.py

```diff
@@ -746,3 +755,506 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
+        """
+        This test is only for checking permission to update basic channel
+        properties like name, description, folder and permission to unarchive
+        the channel. Other things like permission to update group settings and
+        channel privacy are tested separately.
+        """
+        public_stream = self.make_stream("test stream")
+
+        private_stream = self.make_stream("private_stream", invite_only=True)
+        self.subscribe(self.admin, private_stream.name)
+        self.subscribe(self.moderator, private_stream.name)
+        self.subscribe(self.example_user("hamlet"), private_stream.name)
+        self.subscribe(self.example_user("prospero"), private_stream.name)
+
+        unsubscribed_private_stream = self.make_stream(
+            "unsubscribed_private_stream", invite_only=True
+        )
+        # Subscribing a user that is not used in this test, as we do not allow
+        # unarchiving vacant private streams.
+        self.subscribe(self.example_user("desdemona"), unsubscribed_private_stream.name)
+
+        channel_folder = check_add_channel_folder(
+            self.realm, "Frontend", "", acting_user=self.admin
+        )
+
+        for stream in [public_stream, private_stream, unsubscribed_private_stream]:
+            self.do_test_updating_channel(stream, "name", "Renamed stream")
+            self.do_test_updating_channel(stream, "description", "Edited stream description")
+            self.do_test_updating_channel(stream, "folder_id", channel_folder.id)
+
+            do_deactivate_stream(stream, acting_user=None)
+            self.do_test_updating_channel(stream, "deactivated", False)
+
+    def check_channel_privacy_update(
+        self, user: UserProfile, property_name: str, new_value: bool, error_msg: str | None = None
+    ) -> None:
+        stream = get_stream("test_stream", user.realm)
+        data = {}
+        if property_name == "invite_only":
+            data["is_private"] = orjson.dumps(new_value).decode()
+        else:
+            data[property_name] = orjson.dumps(new_value).decode()
+
+        old_value = getattr(stream, property_name)
+        result = self.api_patch(user, f"/api/v1/streams/{stream.id}", info=data)
+
+        if error_msg is not None:
+            self.assert_json_error(result, error_msg)
+            return
+
+        self.assert_json_success(result)
+        stream.refresh_from_db()
+        self.assertEqual(getattr(stream, property_name), new_value)
+
+        # Reset to original value.
+        setattr(stream, property_name, old_value)
+        stream.save(update_fields=[property_name])
+
+        # Reset history_public_to_subscribers field when stream
+        # is changed from private to public.
+        if not stream.invite_only and not stream.history_public_to_subscribers:
+            stream.history_public_to_subscribers = True
+            stream.save(update_fields=["history_public_to_subscribers"])
```

Comment:
I really don't like this model for doing tests -- the leakage risks that some other state remains wrong are considerable, when we could just let Django test system's transactions handle it. It's much better to do each case in a separate test function for this reason.